### PR TITLE
express-notes-api - Joshua Beh

### PR DESCRIPTION
**Pull request by JoshuaBeh**
express-notes-api gif

### CODE REVIEW

File: express-notes-api/delete.js [hunk 1]
```diff
@@ -0,0 +1,23 @@
+module.exports = function deleteEntry(app, data, path, fs, writeForDelete) {
+  app.delete('/api/notes/:id', (req, res) => {
```
Comment (RobertGardner) [inline express-notes-api/delete.js:2]: It would be best to set up all your endpoints in one file and then leave this file to just do the delete functionality. In general, you want functions to just one thing. This one's doing two things: setting up the endpoint and performing the delete action.

File: express-notes-api/delete.js [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    const id = req.params.id;
+    const selectedNote = data.notes[id];
+    delete data.notes[id];
+    if (Number(id) < 0 || isNaN(id)) {
+      const negativeId = { error: 'id must be a positive integer' };
+      res.status(400).json(negativeId);
+      return;
+    } else if (selectedNote === undefined) {
```
Comment (RobertGardner) [inline express-notes-api/delete.js:10]: Since you return above, the `else` here is redundant. (Though, of course, the `if` is necessary.)

File: express-notes-api/delete.js [hunk 3]
```diff
@@ -0,0 +1,23 @@
+module.exports = function deleteEntry(app, data, path, fs, writeForDelete) {
+  app.delete('/api/notes/:id', (req, res) => {
+    const id = req.params.id;
+    const selectedNote = data.notes[id];
+    delete data.notes[id];
```
Comment (RobertGardner) [inline express-notes-api/delete.js:5]: It's best to do your error checking before you make any changes to the data. That way the data will remain in good shape if you end up rejecting the request. It is good, though, that you don't write out the data if there's a request error!

File: express-notes-api/get.js [hunk 4]
```diff
@@ -0,0 +1,29 @@
+function getId(app, data) {
+  app.get('/api/notes/:id', (req, res) => {
+    // Is this fine to use as well? const id = req.url.split('/')[3];
```
Comment (RobertGardner) [inline express-notes-api/get.js:3]: It can be used in this case, but I wouldn't recommend it. It can be hard to parse that `url` correctly and express does it for you, so you should use what they do. But in this case, the two are identical.

File: express-notes-api/post.js [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+      const notFound = { error: 'content is a required field' };
+      res.status(400).json(notFound);
+      return;
+    }
+    const id = data.nextId;
+    data.notes[id] = req.body;
+    data.notes[id].id = id;
+    data.nextId++;
+    write(path, data);
+    if (!fs.existsSync(path)) {
```
Comment (RobertGardner) [inline express-notes-api/post.js:14]: You are checking if the file exists _after_ you've written to it! Check it before and don't do the write if it doesn't exist.

File: express-notes-api/write.js [hunk 6]
```diff
@@ -0,0 +1,18 @@
+const fs = require('fs');
+
+function write(path, data) {
+  fs.writeFile(path, JSON.stringify(data, null, 2), { flag: 'r+' }, (err) => {
```
Comment (RobertGardner) [inline express-notes-api/write.js:4]: I'm curious why you are using `r+` here? It just opens the file for both reading and writing, but since `writeFile` opens the file, writes, then closes the file, I don't think this is doing anything useful.

**Review by RobertGardner — APPROVED** [review]

File: express-notes-api/write.js (hunk 6 above)
Comment (JoshuaBeh) [reply express-notes-api/write.js:4]: From what I can remember, it was how I stopped `fs.writeFile()` from creating a new file if the path didn't exist. Or it was something involving receiving the 500 error along side the `console.error(err)`.
